src/runtime/policy/policy.ts

```diff
@@ -0,0 +1,270 @@
+import {flatMap} from '../util.js';
+
+export enum PolicyEgressType {
+  Logging = 'Logging',
+  FederatedAggregation = 'FederatedAggregation',
+}
+
+export enum PolicyRetentionMedium {
+  Ram = 'Ram',
+  Disk = 'Disk',
+}
+
+export enum PolicyAllowedUsageType {
+  Any = 'Any',
+  Egress = 'Egress',
+  Join = 'Join',
+}
+
+/**
+ * Definition of a dataflow policy.
+ *
+ * This class is just a thin wrapper around the Policy AST node. There's no
+ * actual functionality other than validating the input is correct.
+ */
+export class Policy {
+  constructor(
+      readonly name: string,
+      readonly targets: PolicyTarget[],
+      readonly configs: PolicyConfig[],
+      private readonly annotationRefs: AnnotationRef[]) {
+    this.validateEgressType();
+  }
+
+  get description(): string | null {
+    const description = this.annotationRefs.find(a => a.name === 'intendedPurpose');
+    return description ? description.params['description'] as string : null;
+  }
+
+  get egressType(): PolicyEgressType {
+    const egressType = this.annotationRefs.find(a => a.name === 'egressType');
+    return PolicyEgressType[egressType.params['type'] as string];
+  }
+
+  get annotations() {
+    const specialAnnotations = ['intendedPurpose', 'egressType'];
+    return this.annotationRefs.filter(annotation => !specialAnnotations.includes(annotation.name));
+  }
+
+  private validateEgressType() {
+    const egressType = this.annotationRefs.find(annotation => annotation.name === 'egressType');
+    if (egressType) {
+      checkValueInEnum(egressType.params['type'], PolicyEgressType);
+    }
+  }
+
+  toManifestString(): string {
+    return [
+      ...this.annotationRefs.map(annotation => annotation.toString()),
+      `policy ${this.name} {`,
+      ...flatMap(this.targets, target => indentLines(target.toManifestString())),
+      ...flatMap(this.configs, config => indentLines(config.toManifestString())),
+      '}',
+    ].join('\n');
+  }
+
+  static fromAstNode(
+      node: AstNode.Policy,
+      buildAnnotationRefs: (ref: AstNode.AnnotationRef[]) => AnnotationRef[]): Policy {
+    const annotationRefs = buildAnnotationRefs(node.annotationRefs);
+
+    checkNamesAreUnique(node.targets.map(target => ({name: target.schemaName})));
+    const targets = node.targets.map(target => PolicyTarget.fromAstNode(target, buildAnnotationRefs));
+
+    checkNamesAreUnique(node.configs);
+    const configs = node.configs.map(config => PolicyConfig.fromAstNode(config));
+
+    return new Policy(node.name, targets, configs, annotationRefs);
+  }
+}
+
+class PolicyTarget {
+  constructor(
+      readonly schemaName: string,
+      readonly fields: PolicyField[],
+      private readonly annotationRefs: AnnotationRef[]) {
+    this.validateRetentions();
+    // TODO(b/157605585): Validate maxAge annotation.
+  }
+
+  get retentions() {
+    return this.annotationRefs
+        .filter(annotation => annotation.name === 'allowedRetention')
```

Comment:
maybe worth declaring static const strings for the keywords? probably this is a more java/kt way of doing stuff, but i like it in general.

Reply:
Done